src/host/_stream.cpp

```diff
@@ -179,43 +187,78 @@ void WriteCharsLegacy(SCREEN_INFORMATION& screenInfo, const std::wstring_view& t
     // If it's not set, we can just straight up give everything to WriteCharsLegacyUnprocessed.
     if (WI_IsFlagClear(screenInfo.OutputMode, ENABLE_PROCESSED_OUTPUT))
     {
-        _writeCharsLegacyUnprocessed(screenInfo, { it, end }, psScrollY);
-        it = end;
+        const auto lastCharWrapped = _writeCharsLegacyUnprocessed(screenInfo, text, psScrollY);
+
+        if (io)
+        {
+            // We're asked to produce VT output, but also to behave as if these control characters aren't control characters.
+            // So, to make it work, we simply replace all the control characters with whitespace.
+            io->WriteUTF16StripControlChars(text);
+            if (lastCharWrapped)
+            {
+                io->WriteUTF8(" \r");
+            }

```

Comment:
> I can't remember the rules for scrolling when not in VT mode

When the buffer scrolls, new rows get initialized with the current attributes.

I wonder if I should simply translate these by default to CNL instead. Conhost traditionally had no line wrapping when resizing the window after all.
I'd only translate them to `" \r"` (plus appropriate SGR to set the current attributes) when the cursor is at the end of the last row of the buffer contents that have been written so far. This would at least make line wrapping work when printing (plain) regular text.

I don't mind replicating the console API in VT accurately of course (by reading from the buffer as you suggested), but I'm worried that doing so may run us into similar edge cases around Unicode correctness.

Reply:
> When the buffer scrolls, new rows get initialized with the current attributes.

No, it's weird. Initially new rows are initialized with default attributes (i.e. no color), but once you reach the bottom of the buffer (i.e. when it starts cycling), that's when it begins filling with the current attributes (including color). Of course if you've got no scrollback, then it may look like it's always filling with the current attributes.

Also, even when it's using the current attributes, that's not actually the same as VT. First off the VT fill attributes are dependent on erase color mode (`DECECM`), but even when VT fills with color, it doesn't apply other attributes. So red-on-blue reversed will fill with a blue background in VT mode (i.e. not reversed), but with a red background in the legacy mode.

I don't know how much of that behavior is intentional, but I think it's always worked like that (at least in the v2 console). If it were up to me I'd be inclined to simplify it to always fill new rows with the default attributes, which is at least directly equivalent to VT with `DECECM` enabled.

> I wonder if I should simply translate these to CNL by default instead. Conhost traditionally had no line wrapping when resizing the window after all.

I don't know about anyone else, but I'd be totally fine with that. Although I'm assuming that would have to be something like `\r\n` rather than `CNL`, because `CNL` won't trigger a scroll.

Reply:
> If it were up to me I'd be inclined to simplify it to always fill new rows with the default attributes, which is at least directly equivalent to VT with `DECECM` enabled.

`color` calls `FillConsoleOutputAttribute` followed by `SetConsoleTextAttribute` and then expects that cycled rows get re-initialized with those attributes. I really like this idea, but I suppose this means we'd need to shim the `color` command. I'd be okay with that.

> I don't know about anyone else, but I'd be totally fine with that. Although I'm assuming that would have to be something like `\r\n` rather than `CNL`, because `CNL` won't trigger a scroll.

I'm glad to hear that, and thanks for the suggestion! I might just go with that in this initial PR and refine it if needed in future PRs until 1.22 gets released.